File: app/models/dashboard_stats.rb
# frozen_string_literal: true

# DashboardStats - Read-only model for accessing the dashboard_stats materialized view
#
# This model provides fast access to pre-aggregated dashboard statistics.
# The underlying materialized view is refreshed asynchronously after contact changes.
#
# Usage:
#   stats = DashboardStats.current
#   puts stats.total_contacts
#   puts stats.pending_count
#   puts stats.avg_quality_score
#
class DashboardStats < ApplicationRecord
  self.table_name = 'dashboard_stats'
  self.primary_key = 'updated_at'

  # This is a read-only model - prevent writes
  def readonly?
    true
  end

  # Get the current (latest) stats
  # Since there's only one row, this returns that row
  def self.current
    first || new(total_contacts: 0)
  end

  # Manually refresh the materialized view.
  # Uses a non-concurrent refresh if already inside a transaction.
  def self.refresh!
    if connection.open_transactions.positive?
      connection.execute('REFRESH MATERIALIZED VIEW dashboard_stats;')
    else
      connection.execute('REFRESH MATERIALIZED VIEW CONCURRENTLY dashboard_stats;')
    end
  end

  # Get stats as a hash for easier use in controllers/views
  def to_stats_hash
    {
      # Status breakdown
      pending: pending_count,
      processing: processing_count,
      completed: completed_count,
      failed: failed_count,

      # Phone validation
      valid_numbers: valid_numbers_count,
      invalid_numbers: invalid_numbers_count,

      # Line types
      mobile: mobile_count,
      landline: landline_count,
      voip: voip_count,

      # Business
      businesses: business_count,
      business_enriched: business_enriched_count,

      # Email
      has_email: has_email_count,
      verified_emails: verified_email_count,
      email_enriched: email_enriched_count,

      # Address
      address_enriched: address_enriched_count,

      # Verizon
      verizon_5g_available: verizon_5g_available_count,
      verizon_lte_available: verizon_lte_available_count,
      verizon_checked: verizon_checked_count,

      # Trust Hub
      trust_hub_verified: trust_hub_verified_count,
      trust_hub_enriched: trust_hub_enriched_count,

      # Risk levels
      low_risk: low_risk_count,
      medium_risk: medium_risk_count,
      high_risk: high_risk_count,

      # Duplicates
      duplicates: duplicate_count,

      # Quality
      avg_quality_score: avg_quality_score&.round(2),
      avg_completeness: avg_completeness&.round(2),

      # CRM
      salesforce_synced: salesforce_synced_count,
      hubspot_synced: hubspot_synced_count,
      pipedrive_synced: pipedrive_synced_count,

      # Engagement
      contacted_via_sms: contacted_via_sms_count,
      contacted_via_voice: contacted_via_voice_count,
      sms_opt_outs: sms_opt_out_count,

      # Total
      total_contacts: total_contacts,

      # Metadata
      last_updated: updated_at
    }
  end

  # Calculate percentages for common metrics
  def enrichment_percentage
    return 0 if total_contacts.zero?

    ((business_enriched_count + email_enriched_count + address_enriched_count).to_f / (total_contacts * 3) * 100).round(1)
  end

  def completion_rate
    return 0 if total_contacts.zero?

    (completed_count.to_f / total_contacts * 100).round(1)
  end

  def business_percentage
    return 0 if total_contacts.zero?

    (business_count.to_f / total_contacts * 100).round(1)
  end

  def email_verification_rate
    return 0 if has_email_count.zero?

    (verified_email_count.to_f / has_email_count * 100).round(1)
  end
end

File: laravel-app/app/Models/DashboardStats.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

/**
 * DashboardStats - Read-only model for accessing the dashboard_stats materialized view
 *
 * This model provides fast access to pre-aggregated dashboard statistics.
 * The underlying materialized view would need to be created separately via migration.
 * 
 * Note: In Laravel, we typically use database views or direct queries for this functionality.
 * This is a placeholder model showing the Rails equivalent structure.
 */
class DashboardStats extends Model
{
    /**
     * The table associated with the model.
     *
     * @var string
     */
    protected $table = 'dashboard_stats';

    /**
     * The primary key for the model.
     *
     * @var string
     */
    protected $primaryKey = 'updated_at';

    /**
     * Indicates if the IDs are auto-incrementing.
     *
     * @var bool
     */
    public $incrementing = false;

    /**
     * The attributes that should be cast.
     *
     * @var array<string, string>
     */
    protected $casts = [
        'updated_at' => 'datetime',
        'avg_quality_score' => 'float',
        'avg_completeness' => 'float',
    ];

    /**
     * This is a read-only model - prevent writes
     */
    public function isReadOnly(): bool
    {
        return true;
    }

    /**
     * Get the current (latest) stats
     */
    public static function current(): self
    {
        return self::first() ?? new self(['total_contacts' => 0]);
    }

    /**
     * Get stats as an array for easier use in controllers/views
     */
    public function toStatsArray(): array
    {
        return [
            // Status breakdown
            'pending' => $this->pending_count ?? 0,
            'processing' => $this->processing_count ?? 0,
            'completed' => $this->completed_count ?? 0,
            'failed' => $this->failed_count ?? 0,

            // Phone validation
            'valid_numbers' => $this->valid_numbers_count ?? 0,
            'invalid_numbers' => $this->invalid_numbers_count ?? 0,

            // Line types
            'mobile' => $this->mobile_count ?? 0,
            'landline' => $this->landline_count ?? 0,
            'voip' => $this->voip_count ?? 0,

            // Business
            'businesses' => $this->business_count ?? 0,
            'business_enriched' => $this->business_enriched_count ?? 0,

            // Email
            'has_email' => $this->has_email_count ?? 0,
            'verified_emails' => $this->verified_email_count ?? 0,
            'email_enriched' => $this->email_enriched_count ?? 0,

            // Address
            'address_enriched' => $this->address_enriched_count ?? 0,

            // Verizon
            'verizon_5g_available' => $this->verizon_5g_available_count ?? 0,
            'verizon_lte_available' => $this->verizon_lte_available_count ?? 0,
            'verizon_checked' => $this->verizon_checked_count ?? 0,

            // Trust Hub
            'trust_hub_verified' => $this->trust_hub_verified_count ?? 0,
            'trust_hub_enriched' => $this->trust_hub_enriched_count ?? 0,

            // Risk levels
            'low_risk' => $this->low_risk_count ?? 0,
            'medium_risk' => $this->medium_risk_count ?? 0,
            'high_risk' => $this->high_risk_count ?? 0,

            // Duplicates
            'duplicates' => $this->duplicate_count ?? 0,

            // Quality
            'avg_quality_score' => round($this->avg_quality_score ?? 0, 2),
            'avg_completeness' => round($this->avg_completeness ?? 0, 2),

            // CRM
            'salesforce_synced' => $this->salesforce_synced_count ?? 0,
            'hubspot_synced' => $this->hubspot_synced_count ?? 0,
            'pipedrive_synced' => $this->pipedrive_synced_count ?? 0,

            // Engagement
            'contacted_via_sms' => $this->contacted_via_sms_count ?? 0,
            'contacted_via_voice' => $this->contacted_via_voice_count ?? 0,
            'sms_opt_outs' => $this->sms_opt_out_count ?? 0,

            // Total
            'total_contacts' => $this->total_contacts ?? 0,

            // Metadata
            'last_updated' => $this->updated_at,
        ];
    }

    /**
     * Calculate enrichment percentage
     */
    public function enrichmentPercentage(): float
    {
        if (($this->total_contacts ?? 0) === 0) {
            return 0;
        }

        $enriched = ($this->business_enriched_count ?? 0) + 
                   ($this->email_enriched_count ?? 0) + 
                   ($this->address_enriched_count ?? 0);
        $total = $this->total_contacts * 3;

        return round(($enriched / $total) * 100, 1);
    }

    /**
     * Calculate completion rate
     */
    public function completionRate(): float
    {
        if (($this->total_contacts ?? 0) === 0) {
            return 0;
        }

        return round((($this->completed_count ?? 0) / $this->total_contacts) * 100, 1);
    }

    /**
     * Calculate business percentage
     */
    public function businessPercentage(): float
    {
        if (($this->total_contacts ?? 0) === 0) {
            return 0;
        }

        return round((($this->business_count ?? 0) / $this->total_contacts) * 100, 1);
    }

    /**
     * Calculate email verification rate
     */
    public function emailVerificationRate(): float
    {
        if (($this->has_email_count ?? 0) === 0) {
            return 0;
        }

        return round((($this->verified_email_count ?? 0) / $this->has_email_count) * 100, 1);
    }
}
